mars/dataframe/datasource/from_records.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright 1999-2018 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import pandas as pd
+
+from ...serialize import BoolField, ListField, Int32Field
+from ... import opcodes as OperandDef
+from ..operands import DataFrameOperand, DataFrameOperandMixin, ObjectType
+from ..utils import parse_index
+from ...tensor.core import TENSOR_TYPE
+
+
+class DataFrameFromRecords(DataFrameOperand, DataFrameOperandMixin):
+    _op_type_ = OperandDef.DATAFRAME_FROM_RECORDS
+
+    _columns = ListField('columns')
+    _exclude = ListField('exclude')
+    _coerce_float = BoolField('coerce_float')
+    _nrows = Int32Field('nrows')
+
+    def __init__(self, index=None, columns=None, exclude=None, coerce_float=False, nrows=None,
+                 gpu=False, sparse=False, **kw):
+        if index is not None or columns is not None:
+            raise NotImplementedError('Specifying index value is not supported for now')
+        super(DataFrameFromRecords, self).__init__(_exclude=exclude, _columns=columns,
+                                                   _coerce_float=coerce_float, _nrows=nrows,
+                                                   _gpu=gpu, _sparse=sparse,
+                                                   _object_type=ObjectType.dataframe, **kw)
+
+    @property
+    def columns(self):
+        return self._columns
+
+    @property
+    def exclude(self):
+        return self._exclude
+
+    @property
+    def coerce_float(self):
+        return self._coerce_float
+
+    @property
+    def nrows(self):
+        return self._nrows
+
+    def __call__(self, data):
+        if self.nrows is None:
+            nrows = data.shape[0]
+        else:
+            nrows = self.nrows
+        index_value = parse_index(pd.RangeIndex(start=0, stop=data.shape[0]))
+        dtypes = pd.Series(dict((k, np.dtype(v)) for k, v in data.dtype.descr))
+        columns_value = parse_index(pd.Index(data.dtype.names), store_data=True)
+        return self.new_dataframe([data], (data.shape[0], len(data.dtype.names)), dtypes=dtypes,
+                                  nrows=nrows, index_value=index_value, columns_value=columns_value)
+
+    def _new_chunks(self, inputs, kws=None, **kw):
```

Comment:
Extra parameters can be passed to `new_dataframes` or `new_chunks` directly, they will be added to `_extra_params` automatically, hence this method is not needed.

Reply:
That won't work. Because the problem is not these chunks have the same key, but the `chunk.op` has the same key. The re-calculation of `chunk.op.key` is done before the processing of `_extra_param`: https://github.com/mars-project/mars/blob/master/mars/core.py#L667

Reply:
Can we pass the extra params to op to make the op different?

Reply:
How about add two parameter `begin_index` and `end_index` to `DataFrameFromRecords`? And construct new op rather than `op.copy().reset_key()` here?

Reply:
Actually you can copy, then set the property, because till now, the key is not generated yet.

```
chunk_op = op.copy().reset_key()
chunk_op.extra_params['begin_index'] = xx
```

Reply:
I see, thanks!

Didn't realize the `extra_params` can be set explicitly.

Reply:
Fixed. Note that I have added a `deepcopy` for the `extra_params` in `Operand.copy` in this PR.

Reply:
OK, that makes sense.